### ISSUE TITLE
Training NER: how much text around entities should I supply?

### ISSUE THREAD

**Issue by avissens** [explosion/spaCy#2408]
Hi,
I'm gathering examples to train SpaCy but I noticed that the model gives me different results when I feed it the whole text or just one sentence when it got the entity wrong/right. Below just a couple of examples. To reproduce the same results I need to add a lot of text before/after or sometimes the whole original text. 
I understand why it is happening but my question is - does it matter how much text I supply to the model when training? Can I supply only one short sentence or should I supply enough text around entities to reproduce the result?

#text = "For any British person to have a legal right to a coat of arms it must have been granted to them or they must be descended in the male line from a person to whom arms were awarded. \
#Organisations can also be granted a coat of arms. \
#Coats of arms date back to 12th Century and were traditionally worn over armour in tournaments so participants could identify their opponents." # picks Coat as PERSON
#text = "Coats of arms date back to 12th Century and were traditionally worn over armour in tournaments so participants could identify their opponents." # doesn't pick Coat as PERSON
#text = "Add some text here to reproduce the result. Celotex tested and sold the insulation could amount to corporate manslaughter. Celotex said it could not comment" # picks Celotex as PERSON
#text = "Celotex tested and sold the insulation could amount to corporate manslaughter. Celotex said it could not comment" # doesn't pick Celotex as PERSON

```python
all_tags = nlp(text)
person_list=[]
for ent in all_tags.ents:
    if ent.label_=="PERSON":
        person_list.append(str(ent))      
print (person_list)
```

***

spaCy version: 2.0.11
Platform: Darwin-16.7.0-x86_64-i386-64bit
Python version: 3.6.1
Models: en_core_web_sm

**Comment by honnibal** [comment]
I'm surprised it's so volatile in these examples, since the main features the NER relies on are drawn from a window of four words around the target word. So, in theory a fairly narrow context (individual sentences) should suffice.

One complication is that if you don't provide enough *non-entity* text, the priors around what's an entity will be quite different. If your entity density is low in your test data, but every sentence in your training data has an entity, you're going to have reduced accuracy.

In practice annotating text that doesn't contain entities should be very quick, so I think having enough non-entity text is a good idea.

**Comment by avissens** [comment]
Thank you for your reply! I'm interested only in PERSON entity so in my case the data is sparse indeed. My following question is - should I train the model to recognise all common entities or I can focus only on PERSON?

Another thing that I've noticed is different results when feeding the model the whole text or sentence by sentence. It seems like when it processes single sentences the accuracy is better. I read some posts in this forum about it but didn't find a clear answer. Can this be caused by the way the model is re-trained (using short sentences)? Below are some examples:

```
#text = "Some text before. Common clothes moths and case-bearing clothes moths don't travel very far."
#text = 'Happened to mine from the factory shops, again all my fault! Ours did this a few years ago.'
#text = "According to Andrej Petrovski's analysis, the dozen websites are linked as they are hosted on the same server, which he says means they are shared by the same person, group or company. Some of the sites have had Google Ads, making money when people clicked on them. A whistleblower who knows about Mr Dowson's operation has shown us evidence that these sites made over £100,000 in just an 18-month period."
#text = "Mrs May said the UK should have done more to reduce the risk of the pair being mistreated. On behalf of Her Majesty's government, I apologise unreservedly."
#text = "Add some text here to reproduce the result. Celotex tested and sold the insulation could amount to corporate manslaughter. Celotex said it could not comment"
#text = "Thank you @stellamccartney ! Realized Friday morning the beige dress I was planning to wear to Royal ceremony would photograph too “white” for a wedding.Her team did this overnight. Hat is vintage @philiptreacy been in my closet since 2005 with new feathers. OMG was this an extraordinary day! #Harry&Meghan #RoyalWedding #Lovedeveryminute"
#text = 'The resignation of Amber Rudd is all over the front pages. The Daily Mail says she was "forced" to go because of the pressure on her, adding that her departure is a "huge blow" to Theresa May. The Daily Mirror uses the headline "Good Ruddance", and the Daily Express says the prime minister has been left "in crisis". Meanwhile, the Financial Times reports that supermarkets Sainsburys and Asda will outdo Tesco in terms of market share once they merge. And the Daily Star says TV presenter Ant McPartlin is planning a return to screens in the autumn.'
#text = "It's common for students to take a bit of time out to travel, do some volunteering or grab some extra cash for the years of studying ahead. And who doesn't want to know what a royal would do with their break? Prince William chose to take a breather before starting his degree at St Andrew's University, where he met the duchess. Before their love story began, the duke's year included completing preparation for survival exercises with the Welsh Guards in Belize, central America and working on a British dairy farm."
#text = "For any British person to have a legal right to a coat of arms it must have been granted to them or they must be descended in the male line from a person to whom arms were awarded. Organisations can also be granted a coat of arms. Coats of arms date back to 12th Century and were traditionally worn over armour in tournaments so participants could identify their opponents." 

tokens = nlp(text)

person_list=[]
for ent in tokens.ents:
    if ent.label_=="PERSON":
        person_list.append(str(ent))      
print (person_list)

person_list_sent=[]
sentences = [sent.string.strip() for sent in tokens.sents]
for sent in sentences:
    all_tags = nlp(sent)
    for ent in all_tags.ents:
        if ent.label_=="PERSON":
            person_list_sent.append(str(ent))
print(person_list_sent)
```

**Comment by Rahul6818** [comment]
how much training data per new entity one should supply to add new entities to the existing English models

**Comment by honnibal** [comment]
It's generally hard to tell how many examples will be enough, since it depends on how easy the examples are to predict. You should try the v2.1.0a9 version which will soon be released on `spacy-nightly`, though. It has several fixes to improve adding of entity types to pre-trained models. Thanks for your patience on this.
